prov/util/src/rocr_mem_monitor.c

```diff
@@ -0,0 +1,378 @@
+			"Failed to perform hsa_amd_pointer_info: %s\n",
+			ofi_hsa_status_to_string(hsa_ret));
+
+		ret = -FI_EIO;
+		goto err_free_entry;
+	}
+
+	/* Requested memory length must be within the allocated length. If not,
+	 * user may be trying to monitor unallocated memory.
+	 */
+	if ((uintptr_t) addr + len >
+	    (uintptr_t) hsa_info.agentBaseAddress + hsa_info.sizeInBytes) {
+		FI_WARN(&core_prov, FI_LOG_MR,
+			"Request ROCR buffer length %lu exceeded the actual "
+			"length %lu. Cannot monitor this memory region.\n", len,
+			hsa_info.sizeInBytes);
+		ret = -FI_EFAULT;
+		goto err_free_entry;
+	}
+
+	FI_DBG(&core_prov, FI_LOG_MR,
+	       "ROCR buffer address %p length %lu expanded to %p length %lu\n",
+	       addr, len, hsa_info.agentBaseAddress, hsa_info.sizeInBytes);
+
+	(*entry)->iov.iov_base = hsa_info.agentBaseAddress;
+	(*entry)->iov.iov_len = hsa_info.sizeInBytes;
+
+	ret = ofi_rbmap_insert(&rocr_mm.entries, (void *) &(*entry)->iov,
+			       (void *) *entry, &(*entry)->node);
+	if (ret) {
+		FI_WARN(&core_prov, FI_LOG_MR,
+			"Failed to insert into RB tree: %s\n", strerror(ret));
+		goto err_free_entry;
+	}
+
+	/* Register a deallocation callback for this ROCR memory region. */
+	hsa_ret = ofi_hsa_amd_register_deallocation_callback(hsa_info.agentBaseAddress,
+							     rocr_mm_dealloc_cb,
+							     (void *) hsa_info.sizeInBytes);
+	if (hsa_ret != HSA_STATUS_SUCCESS) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"Failed to perform hsa_amd_register_deallocation_callback: %s\n",
+			ofi_hsa_status_to_string(hsa_ret));
+
+		ret = -FI_EIO;
+		goto err_rbmap_delete_entry;
+	}
+
+	dlist_insert_tail(&(*entry)->entry, &rocr_mm.entry_list);
+
+	FI_DBG(&core_prov, FI_LOG_MR,
+	       "ROCR buffer address %p length %lu monitor entry allocated\n",
+	       hsa_info.agentBaseAddress, hsa_info.sizeInBytes);
+
+	return FI_SUCCESS;
+
+err_rbmap_delete_entry:
+	ofi_rbmap_delete(&rocr_mm.entries, (*entry)->node);
+err_free_entry:
+	free(*entry);
+err:
+	*entry = NULL;
+	return ret;
+}
+
+static void rocr_mm_init(struct ofi_mem_monitor *monitor)
+{
+	dlist_init(&rocr_mm.entry_list);
+	ofi_rbmap_init(&rocr_mm.entries, rocr_rbmap_compare);
+	ofi_monitor_init(&rocr_mm.mm);
+}
+
+static void rocr_mm_cleanup(struct ofi_mem_monitor *monitor)
+{
+	assert(dlist_empty(&rocr_mm.entry_list));
+	ofi_monitor_cleanup(&rocr_mm.mm);
+}
+
+static int rocr_mm_start(struct ofi_mem_monitor *monitor)
+{
+	/* no-op */
+	return FI_SUCCESS;
+}
+
+static void rocr_mm_stop(struct ofi_mem_monitor *monitor)
+{
+	struct rocr_mm_entry *entry;
+
+	while (!dlist_empty(&rocr_mm.entry_list)) {
+		dlist_pop_front(&rocr_mm.entry_list, struct rocr_mm_entry,
+				entry, entry);
+		rocr_mm_entry_free(entry);
+	}
+}
+
+static void rocr_mm_unsubscribe(struct ofi_mem_monitor *monitor,
+				const void *addr, size_t len,
+				union ofi_mr_hmem_info *hmem_info)
+{
+	struct rocr_mm_entry *entry;
+
+	entry = rocr_mm_entry_find(addr, len);
+	if (entry) {
+		rocr_mm_entry_free(entry);
```

Comment:
See below.  Since multiple cached MRs may correspond to the same device region, you can't free the device region until all cached MRs have been released.

Reply:
I'll add a notify within the unsubscribe.